### DISCUSSION TITLE
How can I use left join with JSONB object?

### DISCUSSION THREAD

**Discussion by orenlande** [supabase/supabase#16119] [Questions] [answered]
Hello,

Say I have two tables:
```
Decks
|deck_id   |cards_json   
| 1        | [{card_id: 20, ...}, {card_id:30, ...}]
| 2        | [{card_id: 45, ...}, {card_id:98, ...}]
```

```
Cards
|card id   | ... | ...
|20        | ... | ...
|34        | ... | ...
```

How can I use one query so:
1. I fetch all the row data in decks (`*`)
2. adding to each row list of cards (`cards(*)`) that matches each object in the cards list provided in `deck` table?

The end result should looks like this:
```
[{
  "deck_id": 1,
   ... more data related to deck ...,
  "cards": [
    {"card_id": 20, ...},
    {"card_id": 30, ...}
  ]
}]
```

Thanks for help!

**Comment by GaryAustin1** [comment]
The API uses foreign keys for joins. You would need to have your cards have a foreign key to decks or have a join table with both cards and decks foreign keys if cards can be in multiple decks.  

Otherwise you would need to attempt it in SQL with an rpc call to a Postgres function.

**Reply by orenlande** [reply]
@GaryAustin1 Thanks for your quick answer!

I Indeed suspected that I might need rpc call to get it done in the way I looked for.

In term of speed and cost-efficient solutions, it indeed would be better to make a new table that links the cards to decks. (aka `deck_cards` table with foreign keys to decks and cards). 

However, In such a case, I'd still need to pass through a third table to link them all, resulting in two queries (one to fetch all card ids related to decks and the other one to fetch all cards data I originally fetched from linked deck)

I'm bit new to postgres and supabase. What's the right solution here? Do you have any ideas on how to make such a connection efficient?

**Reply by GaryAustin1** [reply, marked as answer]
Not sure why you need two queries.  If you have a join table with two fk's, make sure you add them to the primary key (part of a newer requirement for the API to find them).   Then you can just do `.select('*, cards(*)')` on the decks table. The API will find the relationship.
![image](https://github.com/supabase/supabase/assets/54564956/d81415c2-7aaf-4a26-86b4-132d9077f40d)
